docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+waking up the topic.  This should not create a thundering herd during
+node evacuation, because partition movement is already scheduled to 
+avoid moving all at once.
+
+Any continuous data rebalancing feature that we implement should
+prefer to avoid migrating hibernated topics, although it can wake
+them up if they're the only viable candidates to resolve resource
+imbalances.
+
+All internal partitions will be exempt from hibernation.
+
+Partition health monitor data will include an extra field for
+last access time of a partition, or perhaps just the most-recent
+access per topic on each node, so that the controller can use this
+data to figure out the global per-topic access time to enforce
+hibernation timeout.
+
+## Telemetry & Observability
+
+- Gauge for number of topics hibernated
+- Gauge for number of partitions hibernated
+- Counter for number of hibernate events
+- Counter for number of wake events
+- rpk commands for listing topics by status
+
+## Detailed design
+
+### List of controller commands
+### List of RPCs
+
+## Corner cases dissected by example.
+
+## Drawbacks
+
+### Resource determinism & what happens if all partitions go active?
+
+Redpanda's resource footprint
+will grow and shrink depending on the number of partitions that are
+idle.  We can mitigate that somewhat if we set a ceiling on the number
+of partitions that may be active, and force-sleep the least recently used
+partitions when we hit the limit, but that doesn't help if all the partitions
+in the system really are active (we'd just be flapping them in and out of
+hibernation).
+
+That nondeterministic footprint means this only makes sense at some
+statistical scale where the ratio of idle partitions can be proactively
+monitored and the system grown/shrunk in response.  For example, we might
+have a system that is in principle able to support 1M partitions across
+ten nodes, but when many are idle it might shrink right down to a 3 node
+cluster.  We would re-expand the cluster (either by hand or via some
+external infrastructure automation) if that idle ratio changed.
+
+### Latency to wake
+
+There will be a short delay on the first I/O to a partition that wakes
+it from hibernation.  On a healthy system this should be in the range
+of a few hundred milliseconds.
+
+## Alternatives
+
+### Do nothing
+
+The practical capacity of a system will be how many _active_ partitions
+we can handle: this has the advantage of determinism, as there is no
+risk of a system 
+
+### Radically improved resource management elsewhere
+
+Separately to this RFC, we know that various subsystems' resource management
+can be improved, such as the way we keep unbounded in-memory indices in
+the storage layer, and hold open file descriptors for segments.
+
+In the raft layer, hibernation would be less important if we moved to a
+model where several partitions were grouped together in one raft group,
+thereby avoiding the "heartbeat per partition" scaling issue.  This would
```

Comment:
This heartbeat issue seems to keep coming up as a scaling problem, are we tracking it anywhere to be solved?